lib/std/array_list.zig

```diff
@@ -633,6 +633,15 @@ pub fn ArrayListAlignedUnmanaged(comptime T: type, comptime alignment: ?u29) typ
             return self;
         }
 
+        /// Initialize with externally-managed memory. The buffer determines the
+        /// capacity, and the length is set to zero.
+        pub fn initBuffer(buffer: []T) Self {
+            return .{
+                .items = buffer[0..0],
+                .capacity = buffer.len,
+            };
+        }
```

Comment:
Note that the resulting `ArrayListUnmanaged` will be full of footguns, as any function that can resize the list will try to resize or free the current `items` slice, which will either be illegal behavior (freeing a pointer to a stack-allocated array) or will likely lead to a double free later on (since the caller will likely free the `buffer` as well).

Would it make sense to have something like a `BufferList` type for this use-case that can't resize but still has the convenience functions of `ArrayList` (append, appendSlice, Writer, etc)? Would probably end up being pretty similar to `BoundedArray` but would take a slice instead of having an array as a field.

Reply:
Maybe it's enough to add to add sentence to the doc comment; something like 'It is undefined behavior to call any method with an allocator parameter on the result, use the `*AssumeCapacity` variants.'

Reply:
Here's what happens in safe build modes:

```zig
const std = @import("std");

test "invalid free" {
    var buffer: [1]u8 = undefined;
    var list = std.ArrayListUnmanaged(u8).initBuffer(&buffer);

    try list.append(std.testing.allocator, 100);
    try list.append(std.testing.allocator, 200);
}
```

```
Test [1/1] test.invalid free... thread 1811364 panic: Invalid free
/home/andy/dev/zig/lib/std/heap/general_purpose_allocator.zig:567:21: 0x22654f in resizeLarge (test)
                    @panic("Invalid free");
                    ^
/home/andy/dev/zig/lib/std/heap/general_purpose_allocator.zig:735:40: 0x223264 in resize (test)
                return self.resizeLarge(old_mem, log2_old_align, new_size, ret_addr);
                                       ^
/home/andy/dev/zig/lib/std/mem/Allocator.zig:92:30: 0x270f7b in resize__anon_8229 (test)
    return self.vtable.resize(self.ptr, buf, log2_buf_align, new_len, ret_addr);
                             ^
/home/andy/dev/zig/lib/std/array_list.zig:1013:33: 0x2348f4 in ensureTotalCapacityPrecise (test)
            if (allocator.resize(old_memory, new_capacity)) {
                                ^
/home/andy/dev/zig/lib/std/array_list.zig:992:51: 0x22af88 in ensureTotalCapacity (test)
            return self.ensureTotalCapacityPrecise(allocator, better_capacity);
                                                  ^
/home/andy/dev/zig/lib/std/array_list.zig:1045:41: 0x227c7b in addOne (test)
            try self.ensureTotalCapacity(allocator, newlen);
                                        ^
/home/andy/dev/zig/lib/std/array_list.zig:804:49: 0x222d95 in append (test)
            const new_item_ptr = try self.addOne(allocator);
                                                ^
/home/andy/dev/zig/lib/test.zig:8:20: 0x222cab in test.invalid free (test)
```

I think it's fine.

Reply:
Just a note: this should be `Slice` instead of `[]T`.